File: app/core/rlhf.py
import os
import json
import uuid
from datetime import datetime, timezone

DATA_DIR = os.path.join(os.getcwd(), "data", "rlhf")

def _ensure_dir():
    os.makedirs(DATA_DIR, exist_ok=True)

def log_feedback(intent: str, model: str, rating: int, feedback_text: str = "") -> dict:
    """"Log user feedback (RLHF) to a persistent store."""
    _ensure_dir()
    path = os.path.join(DATA_DIR, "feedback.json")
    records = []
    if os.path.exists(path):
        with open(path, "r", encoding="utf-8") as f:
            try:
                records = json.load(f)
            except json.JSONDecodeError:
                pass
    record = {
        "id": str(uuid.uuid4()),
        "intent": intent,
        "model": model,
        "rating": rating,
        "feedback_text": feedback_text,
        "created_at": datetime.now(timezone.utc).isoformat()
    }
    records.append(record)
    with open(path, "w", encoding="utf-8") as f:
        json.dump(records, f, indent=2)
    return record

File: npm/src/core/rlhf.ts
import { randomUUID } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";

export interface FeedbackRecord {
  id: string;
  intent: string;
  model: string;
  rating: number; // 1 to 5
  feedbackText: string;
  createdAt: string;
}

let dataDir = join(process.cwd(), "data", "rlhf");

function ensureDir(): void {
  if (!existsSync(dataDir)) {
    mkdirSync(dataDir, { recursive: true });
  }
}

function persistPath(): string {
  ensureDir();
  return join(dataDir, "feedback.json");
}

export function logFeedback(intent: string, model: string, rating: number, feedbackText: string = ""): FeedbackRecord {
  const p = persistPath();
  let records: FeedbackRecord[] = [];
  if (existsSync(p)) {
    try {
      records = JSON.parse(readFileSync(p, "utf-8"));
    } catch {}
  }
  const full: FeedbackRecord = {
    id: randomUUID(), intent, model, rating, feedbackText, createdAt: new Date().toISOString()
  };
  records.push(full);
  writeFileSync(p, JSON.stringify(records, null, 2), "utf-8");
  return full;
}
